### ISSUE TITLE
Examine reducing overhead by removing the "Controller" structs

### ISSUE THREAD

**Issue by tomaka** [libp2p/rust-libp2p#365]
Right now several part of the library look like this:

```rust
pub struct FooController {
     shared: Arc<Mutex<Shared>>,
}

pub struct FooFuture {
     shared: Arc<Mutex<Shared>>,
}

struct Shared {
    ...
}
```

The reason is that most `Future` operations take ownership of the future. In other words, the `FooFuture` you obtain would be consumed and you'd no longer be able to access it.

However it should also be possible to do this:
```rust
let foo_future = Arc::new(Mutex::new(foo_future));
future::poll_fn({ let f = foo_future.clone(); move || f.lock().poll() })
```

Which is exactly the same, except that now we can put the methods on the `FooFuture` instead of the `FooController`, and the user is responsible for the overhead.

**Comment by tomaka** [comment]
An advantage of the `Controller` part, however, is that the controller sometimes doesn't require a typing in its generics that is as strong as the `Future`.

**Comment by tomaka** [comment]
No longer relevant after #565
